stdlib/public/Concurrency/AsyncStreamBuffer.swift

```diff
@@ -0,0 +1,384 @@
+
+    func cancel() {
+      switch self {
+      case .nonThrowing(let handler):
+        handler(.cancelled)
+      case .throwing(let handler):
+        handler(.cancelled)
+      }
+    }
+
+    func finish(throwing error: Error? = nil) {
+      switch self {
+      case .nonThrowing(let handler):
+        handler(.finished)
+      case .throwing(let handler):
+        handler(.finished(error))
+      }
+    }
+  }
+
+  struct State {
+    // continuation can be either of two types:
+    // UnsafeContinuation<Element?, Error>
+    // UnsafeContinuation<Element?, Never>
+    var continuation: Builtin.RawUnsafeContinuation?
+    var pending = [Element]()
+    var terminal: Terminal?
+    var onTermination: TerminationHandler?
+    let limit: Int
+
+    init(limit: Int) {
+      self.limit = limit
+    }
+  }
+  // Stored as a singular structured assignment for initialization
+  var state: State
+
+  private init(_doNotCallMe: ()) {
+    fatalError("Storage must be initialized by create")
+  }
+
+  deinit {
+    state.onTermination?.cancel()
+  }
+
+  private func lock() {
+    let ptr =
+      UnsafeRawPointer(Builtin.projectTailElems(self, UnsafeRawPointer.self))
+    _lock(ptr)
+  }
+
+  private func unlock() {
+    let ptr =
+      UnsafeRawPointer(Builtin.projectTailElems(self, UnsafeRawPointer.self))
+    _unlock(ptr)
+  }
+
+  func getOnTermination() -> (@Sendable (AsyncStream<Element>.Continuation.Termination) -> Void)? where Failure == Never {
+    lock()
+    let handler = state.onTermination
+    unlock()
+    switch handler {
+    case .nonThrowing(let onTermination):
+      return onTermination
+    default:
+      return nil
+    }
+  }
+
+  func getOnTermination() -> (@Sendable (AsyncThrowingStream<Element>.Continuation.Termination) -> Void)? {
+    lock()
+    let handler = state.onTermination
+    unlock()
+    switch handler {
+    case .throwing(let onTermination):
+      return onTermination
+    default:
+      return nil
+    }
+  }
+
+  func setOnTermination(_ newValue: (@Sendable (AsyncStream<Element>.Continuation.Termination) -> Void)?) where Failure == Never {
+    lock()
+    withExtendedLifetime(state.onTermination) {
+      state.onTermination = newValue.map { .nonThrowing($0) }
+      unlock()
+    }
+  }
+
+  func setOnTermination(_ newValue: (@Sendable (AsyncThrowingStream<Element>.Continuation.Termination) -> Void)?) {
+    lock()
+    withExtendedLifetime(state.onTermination) {
+      state.onTermination = newValue.map { .throwing($0) }
+      unlock()
+    }
+  }
+
+  func cancel() {
+    lock()
+    // swap out the handler before we invoke it to prevent double cancel
+    let handler = state.onTermination
+    state.onTermination = nil
+    unlock()
+    
+    handler?.cancel() // handler must be invoked before yielding nil for termination
+
+    yield(nil)
+  }
+
+  private func enqueue(_ value: __owned Element?) -> TerminationHandler? {
```

Comment:
Could you add a comment on `enqueue` and `terminate` that they can only be called inside a lock?

Reply:
will do.

Reply:
fixed, also while auditing that I caught an issue where the termination handler was called after terminal events were propagated.